design/0001-frr.md

```diff
@@ -0,0 +1,347 @@
+                              │ L2_controller │     │ bgp_controller│
+                              │   <<class>>   │     │   <<class>>   │
+                              │               │     │               │
+                              └───────────────┘     └───────┬───────┘
+                                                            │
+                                                            │  <<use>>
+                                                    ┌───────▼───────┐
+                                                    │               │
+                                                    │    session    │
+                                                    │ <<interface>> │
+                                                    │               │
+                                                    └───────▲───────┘
+                                                            │ <<implements>>
+                                                            │
+                                                    ┌───────┴───────┐
+                                                    │               │
+                                                    │     bgp       │
+                                                    │  <<package>>  │
+                                                    │               │
+                                                    └───────────────┘
+
+```
+
+##### main
+
+The main package contains the main() entrypoint for the speaker application. It
+creates an L2 and/or BGP controller, a speakerList and a k8s client. The k8s
+client watches for changes to the MetalLB config map, service configuration, and
+the node configuration and calls the appropriate protocol handler (e.g. L2, BGP)
+via the “Protocol” Interface after ensuring the change is valid.
+
+The BGP implementation of the “Protocol” interface can be found
+[here](https://github.com/metallb/metallb/blob/main/speaker/bgp_controller.go).
+This handles changes to the config map, service, and node configuration for BGP
+pools. Based on these changes, BGP sessions will be established/destroyed and
+prefixes will be advertised via the bgp package.
+
+##### bgp
+
+The bgp package manages BGP sessions and advertisements. A BGP session is
+created by the New() method and prefixes are advertised by the Set() method
+which is part of the session interface.
+
+BGP sessions are uniquely identified and configured by the parameters of the
+New() method which are currently:
+
+* **addr**: IP address and port of peer
+* **asn**: Local ASN
+* **routerID**: Specify the BGP Router ID of the session
+* **peerASN**: ASN of the router that we are connecting to
+* **holdTime**: BGP Hold Time specifies how long to maintain sessions for
+* **Password**: TCP MD5 sock option for connection
+* **myNode**
+
+Advertisements are made for a session via the Set() method.
+
+##### speakerlist
+
+speakerlist uses the vagrant memberlist package to create a cluster of bgp
```

Comment:
s/vagrant/Hashicorp/

Reply:
> Hi @markdgray
> 
> I went through the document quickly, for me this integration is a great occasion to create a separate BGP controller instead of a keeping the idea of BGP speaker around:

I'm not sure I understand. Are you suggesting creating another Kubernetes controller just to configure BGP? MetalLB would then use that to configure FRR indirectly? I did consider this model but I have some concern about the latency of receiving and handling changes to kubernetes objects by this controller. I am unsure how responsive we would need to be?
> 
>     1. you don't need to run FRR/BGP on each nodes
> 
>     2. you don't need hostNetwork=true
> 
>     3. you don't need the added Linux capabilities
> 
> 
> Having some BGP controllers running only on the control-plane would be great for security
> 
> My 2 cents

Reply:
Not sure to understand the concern about responsiveness, the current speakers are using client-go to talk to the API, such BGP controller would do the same. It's just 2 different models, having all nodes peer to the router is a bit weird to me, it does allow you to run BFD which is nice, but I don't know if it's standard / expected to peer all nodes or not.

Reply:
Ok, I think I see what you are saying. You are suggesting that we don't need all nodes to peer to routers. There could be efficiencies to be gained by only having a subset to peer to routers. That may be possible. I think we would still need to have metallb instances on each node to detect when a node goes down but, potentially, we could not deploy FRR to every node. That would preclude BFD as you mentioned which would be a disadvantage. 

I don't _think_ we could always guarantee there would only be one BGP controller as it may not always be possible to route BGP messages to some routers from a particular node (there could be some administrative constraint). That is why I said "subset" above. If this is the case, it would complicate the design as you would need to figure out which nodes should have an instance of FRR and which shouldn't. Maybe some one else has an opinion?

Reply:
Also, this makes the assumption that the BGP controller can advertise routes with a NEXT_HOP that is not itself.

Reply:
I did a bit of research on this. Typically, a router will host its own BGP stack which will then peer with a BGP agent on another router. In our case, the node is acting as a router and the BGP speaker is acting as the router's BGP stack. Therefore, I think the model currently employed by MetalLB is the more standard approach. Additionally, as you suggested, we will need to deploy FRR (or an alternative) like this in order to make use of BFD. Therefore, I think we should maintain the current model. What do you think?

Reply:
I agree with your conclusion, Mark

Reply:
Thinking about it more, and mostly to have BFD, the current approach is the way to go. Deploying on each node to use memberlist as a BFD replacement just to have less peering makes little sense.
We could just add the option in the Chart to disable L2 and in that case remove the extra capabilities, so people running BGP only can make it a bit more secure, but that's out of scope for this PR

Reply:
Ok, thanks for the review @champtar